packages/agent/src/connect-v2.ts

```diff
@@ -0,0 +1,199 @@
+  const clientDid = await DidDht.create();
+
+  // Hash the code verifier to use as a code challenge and to encrypt the Request Object.
+  const { codeVerifieru8a } = Oidc.generateRandomCodeVerifier();
+
+  // Derive the code challenge based on the code verifier
+  const { codeChallengeu8a, codeChallengeb64url } =
+    await Oidc.deriveCodeChallenge(codeVerifieru8a);
+
+  // build callback URL to pass into the auth request
+  const callbackEndpoint = Oidc.buildOidcUrl({
+    baseURL  : connectServerUrl,
+    endpoint : 'callback',
+  });
+
+  // build the PAR request
+  const request = await Oidc.createAuthRequest({
+    client_id             : callbackEndpoint,
+    scope                 : 'web5', // TODO: clear with frank
+    code_challenge        : codeChallengeb64url,
+    code_challenge_method : 'S256',
+    permission_requests   : permissionRequests,
+    redirect_uri          : callbackEndpoint,
+    // known customer credential defines these
+    client_metadata       : {
+      client_uri                     : clientUri,
+      subject_syntax_types_supported : ['did:dht'],
+    },
+  });
+
+  // Sign the Request Object using the Client DID's signing key.
+  const requestJwt = await Oidc.signJwt({
+    did  : clientDid,
+    data : request,
+  });
+
+  if (!requestJwt) {
+    throw new Error('Unable to sign requestObject');
+  }
+
+  // Encryption nonce
+  const nonce = utils.randomBytes(24);
+
+  // Encrypt the Request Object JWT using the code challenge.
+  const requestObjectJwe = await Oidc.encryptAuthRequest({
+    jwt           : requestJwt,
+    codeChallenge : codeChallengeu8a,
+    nonce,
+  });
+
+  // Convert the encrypted Request Object to URLSearchParams for form encoding.
+  const formEncodedRequest = new URLSearchParams({ request: requestObjectJwe });
+
+  const pushedAuthorizationRequestEndpoint = Oidc.buildOidcUrl({
+    baseURL  : connectServerUrl,
+    endpoint : 'pushedAuthorizationRequest',
+  });
+
+  try {
+    const parResponse = await fetch(pushedAuthorizationRequestEndpoint, {
+      body    : formEncodedRequest,
+      method  : 'POST',
+      headers : {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    });
+    if (!parResponse.ok) {
+      throw new Error(`${parResponse.status}: ${parResponse.statusText}`);
+    }
+
+    const parData: PushedAuthResponse = await parResponse.json();
+
+    // a deeplink to a web5 compatible wallet. if the wallet scans this link it should receive
+    // a route to its web5 connect provider flow and the params of where to fetch the auth request.
+    const walletURI = new URL('web5://connect/');
+    walletURI.searchParams.set(
+      'nonce',
+      Convert.uint8Array(nonce).toBase64Url()
+    );
+    walletURI.searchParams.set('request_uri', parData.request_uri);
+    walletURI.searchParams.set('client_did', clientDid.uri);
+    walletURI.searchParams.set('code_challenge', codeChallengeb64url);
```

Comment:
Sanity: I am under the impression that only `request_uri` and `client_id` are used to form the wallet URI. The rest are fetched directly from the Request URI itself?

Reply:
not the case. the code challenge is necessary.

Reply:
Sanity 2x plus eduction for me: can you explain and/or point me to documentation that says challenge and nonce much be encoded in the URI when they are already part of the Auth Request object? 🙏 

Seems silly to me that the point of Auth Request object is to alleviate the need to pass properties via query params, but you have to do it anyways.

Reply:
> Seems silly to me that the point of Auth Request object is to alleviate the need to pass properties via query params, but you have to do it anyways.

It's a choice that was made to encrypt the auth response with the code challenge because we don't do server side validation like a real IDP would, because we don't have a real server and we aren't a real IDP. In the pure OAuth flow there'd be a way of authenticating the person fetching from the `request_uri`.

Reply:
Keep in mind that in the future there might be transmission mechanisms that don't require an intermediary server at all. We want to future proof for that and not go all in on server side validation. Right now our server is "dumb" and we can keep it that way with this approach.

Reply:
update:
- removed the nonce from the query params
- going to remove `client_did` from the query params

code challenge will remain but can be tabled for a discussion about the future web5 connect spec and whether encrypting with the `code_challenge` was truly the right decision, and we can evolve the spec to be what we want with input from @frankhinek @csuwildcat .

Reply:
should be good to close this out and make a ticket about encrypting with the `code_challenge`

Reply:
Thanks for removing `nonce` from query params.

I see that you also removed `client_id` (your `client_did`) from the URL, though, from examples that I can find in the specs on the internet, I always see `client_id` included in the URL showing that it is an exception to the rule when using `request_uri`. I am not going to insist to include `client_id` in the URL because you are right in saying that it is redundant, but it does "appear unconventional" to not include it. The rationale for having it in the URL seems so that it can act as a request lookup key/short-cut, but that's not how we implement `request_uri` generation.

You are good to resolve this comment thread if you prefer no further action on `client_id`. I'll create a separate thread on the topic of `code_challenge` to not bloat this thread any further!